guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+🚫
+
+```
+src/
+  permission-controller.ts
+test/
+  permission-controller.ts
+```
+
+✅
+
+```
+src/
+  permission-controller.ts
+  permission-controller.test.ts
+```
+
+## Use `describe` to group tests for the module under test
+
+Wrap all tests for a file in a `describe` block named the same as the file itself:
+
+🚫
+
+``` typescript
+// file-utils.ts
+
+export function isFile() {
+  // ...
+}
+
+export function isDirectory() {
+  // ...
+}
+
+// file-utils.test.ts
+
+describe('isFile', () => {
+  // ...
+});
+
+describe('isDirectory', () => {
+  // ...
+});
+```
+
+✅
+
+``` typescript
+// file-utils.ts
+
+export function isFile() {
+  // ...
+}
+
+export function isDirectory() {
+  // ...
+}
+
+// file-utils.test.ts
+
+describe('file-utils', () => {
+  describe('isFile', () => {
+    // ...
+  });
+
+  describe('isDirectory', () => {
+    // ...
+  });
+});
+```
+
+## Use `describe` to group tests for methods
+
+Using `describe` to wrap tests for a method makes it easier to spot these tests in a large file and makes it possible to run them on their own (using `.only`). It also helps to establish the subject of the test (the "it") and therefore keeps the test well named and focused.
+
+🚫
+
+``` typescript
+describe('KeyringController', () => {
+  it('adds a new account to the given keyring', () => {
+    // ...
+  });
+
+  it('removes the account identified by the given address from its associated keyring', () => {
+    // ...
+  });
+});
+```
+
+✅
+
+``` typescript
+describe('KeyringController', () => {
+  describe('addAccount', () => {
+    it('adds a new account to the given keyring', () => {
+      // ...
+    });
+  });
+
+  describe('removeAccount', () => {
+    it('removes the account identified by the given address from its associated keyring', () => {
+      // ...
+    });
+  });
+});
+```
+
+## 💡 Use `describe` to group tests for scenarios
+
+If you have many tests that verify the behavior of a piece of code under a particular condition, you might find it helpful to wrap those tests with a `describe` block. Use a phrase starting with `if ...` or `when ...` so as to form a sentence when combined with the test name.
+
+1️⃣
+
+``` typescript
+it('calls addToken on the tokens controller if "addToken" is true', () => {
+  // ...
+})
+
+it('adds the new token to the state if "addToken" is true', () => {
+  // ...
+})
+
+it('returns a promise that resolves to true if "addToken" is true', () => {
+  // ...
+})
+```
+
+2️⃣
+
+``` typescript
+describe('if "addToken" is true', () => {
+  it('calls addToken on the tokens controller', () => {
+    // ...
+  });
+
+  it('adds the new token to the state', () => {
+    // ...
+  });
+
+  it('returns a promise that resolves to true', () => {
+    // ...
+  });
+});
+```
+
+## Use `it` to specify known behavior for a unit of code
+
+Use `it` to describe the behavior of a piece of code from the consumer's perspective as it stands today, not as it ought to be.
```

Comment:
Hmm. This doesn't seem right. I think test descriptions should describe intentions, i.e. what the code _should_ be doing. Of course they also have to be true (i.e. reflective of what the code actually does), but that's a given. Otherwise the tests wouldn't pass.

We could model the behaviour of the code in many different ways, but organizing by intent produces useful and readable tests.

Reply:
You're right. When I was writing this guideline I was thinking about where I got my current ideas about testing. I have some predispositions from working at consultancies and I remembered that there were two ideas floating around: 1) "tests as specifications" — which I interpret as "tests are valuable for verifying the behavior of a part of the system but also describing that behavior" — and 2) "tests as examples" — which I interpret as "tests are valuable for demonstrating to consumers how to use it".

At the same time, I was trying to think of a better reason for why it would be desirable to remove "should" from test descriptions other than "it's noise". That led me to the question, if tests describe the behavior of the system, and we don't want to use "should", what we are saying in the test descriptions? Are we describing how the system ought to work or how it actually works? I could see how it could be helpful to be honest and admit "here's what we've got today", whether it's desirable or not. This perspective is especially useful, I think, if you're backfilling tests for code that wasn't written with tests to begin with and you're discovering behavior that people may not realize is there.

That's what I was aiming for in writing this guideline, but I'm realizing that it's a nuanced viewpoint and not applicable in general situations.

All this to say, I agree with you that thinking about intended behavior is more important in most situations than thinking about the actual behavior.

So now the question is, if intention is important, do we want to be using "should" in tests? The [original thinking](https://dannorth.net/introducing-bdd/#an-expressive-test-name-is-helpful-when-a-test-fails) behind "should" was that it baked uncertainty into the software development process and served as a reminder to [constantly compare what the code is doing vs. what it ought to be doing (so, kind of the other side of the coin of what I said above). Is this valuable? Or do we already do this anyway?

Reply:
I do like your "tests as specifications" interpretation, that's how I think of them as well. But I would expect a specification to communicate the intention behind each guideline as well (at least for anything non-obvious).

There are some cases where we might be writing tests after-the-fact, and we find something non-desirable, and we want to model it with a unit test anyway just so we don't lose track of the behavior. But that would only happen when backfilling tests, so I think of that as an edge case. "Intention" maybe isn't as relevant if it's something a past author wrote that you don't agree with.

But if it's code you authored to serve a specific purpose, the test description is a great place to state that purpose.

Reply:
I'm still good with omitting "should". I saw it as being better to exclude because the tests then can be read as a specification, and also because it added verbosity to the test descriptions. But I never felt strongly about that guideline either way.

Reply:
Okay, I've updated this section. Hopefully these changes make more sense: 6646934...9f89251